src/sfloader/fluid_instpatch.c

```diff
@@ -0,0 +1,667 @@
+    IpatchDLSFile *file = NULL;
+    IpatchFileHandle *handle = NULL;
+
+    fluid_return_val_if_fail(filename != NULL, NULL);
+
+    if((patchfont = FLUID_NEW(fluid_instpatch_font_t)) == NULL)
+    {
+        return NULL;
+    }
+
+    FLUID_MEMSET(patchfont, 0, sizeof(*patchfont));
+
+    FLUID_STRNCPY(&patchfont->name[0], filename, sizeof(patchfont->name));
+
+    /* open a file, we get a reference */
+    if((file = ipatch_dls_file_new()) == NULL)
+    {
+        FLUID_FREE(patchfont);
+
+        return NULL;
+    }
+
+    /* TODO
+    ipatch_file_set_iofuncs(file, fcbs);
+    */
+
+    /* ipatch_file_open() references the file again */
+    if((handle = ipatch_file_open(IPATCH_FILE(file), filename, "r", &err)) == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "ipatch_file_open() failed with error: '%s'", ipatch_gerror_message(err));
+
+        g_object_unref(file);
+        FLUID_FREE(patchfont);
+
+        return NULL;
+    }
+
+    /* get rid of the reference we own, we dont need it any longer */
+    g_object_unref(file);
+    file = NULL;
+
+    /* open a reader, this gives us a reference */
+    if((reader = ipatch_dls_reader_new(handle)) == NULL)
+    {
+        ipatch_file_close(handle);
+        FLUID_FREE(patchfont);
+
+        return NULL;
+    }
+
+    patchfont->dls = ipatch_dls_reader_load(reader, &err);
+
+    /* unref the reader directly afterwards, not needed any longer */
+    g_object_unref(reader);
+    reader = NULL;
+
+    if(patchfont->dls == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "ipatch_dls_reader_new() failed with error: '%s'", ipatch_gerror_message(err));
+
+        // reader has already been unrefed, i.e. no need to call ipatch_file_close()
+
+        FLUID_FREE(patchfont);
+
+        return NULL;
+    }
+    else
+    {
+        /* at this point everything is owned by the IpatchDLS2*, no need for custom cleanups any longer */
+        
+        IpatchIter iter;
+        IpatchDLS2Inst *inst;
+        gboolean success = ipatch_container_init_iter(IPATCH_CONTAINER(patchfont->dls), &iter, IPATCH_TYPE_DLS2_INST);
+        if(success == FALSE)
+        {
+            goto bad_luck;
+        }
+
+        inst = ipatch_dls2_inst_first(&iter);
+        
+        if(inst == NULL)
+        {
+            FLUID_LOG(FLUID_ERR, "A soundfont file was accepted by libinstpatch, but it doesn't contain a single instrument. Dropping the whole file.");
+            goto bad_luck;
+        }
+
+        /* loop over instruments, convert to sf2 voices, create fluid_samples, cache all presets in a list */
+        do
+        {
+            fluid_preset_t *preset;
+            IpatchSF2VoiceCache* cache;
+            int bank, prog;
+            ipatch_dls2_inst_get_midi_locale(inst, &bank, &prog);
+            
+            if((cache = convert_dls_to_sf2_instrument(patchfont, inst)) == NULL)
+            {
+                FLUID_LOG(FLUID_WARN, "Failed to cache DLS instrument bank %d , prog %d", bank, prog);
```

Comment:
"Failed to **convert** DLS instrument ..."?

Reply:
Yeah, but it's also assigned to a cache, idk.

Reply:
"convert and cache" ? :-)